### PR TITLE
wip: Move to Go modules

### LINKED ISSUE #1331: Migration from 'go dep' to 'go mod'
Coming from [#1152](https://github.com/kata-containers/runtime/pull/1152#issuecomment-470052791):

> [@umarcor]
> @Pennyzct, @jodh-intel just because of curiosity, why are you using `dep` instead of `go mod`?

> [@grahamwhaley]
> @umarcor - at this point, it may be mostly due to historical reasons - the kata, and forerunner, code bases have been around a few years now... I think `go mod` came in with golang 1.11 ?, and we still have our baseline golang supported version as 1.10.4 :-)
> 
> Now, if you'd like to propose a move to `go mod`, and submit a PR ;-) But, first probably best if it was discussed. /cc @sboeuf

> [@jodh-intel]
Yes - at the time we introduced it, `dep` was the best tool available.

> [@sboeuf]
> @umarcor @grahamwhaley @jodh-intel I recently discussed this with @mcastelino  and I think it is time for us to move to go modules. We need to open  issues on every Kata repositories for this. And then let's start submit  PRs. @umarcor is it something you're interested in helping with?

Let's continue here, to avoid getting @Pennyzct's PR dirtier.

# Description of problem

This is neither a problem nor a requirement, because `dep` is not deprecated, yet. Certainly, it is still actively maintained (golang/dep#2052). Nonetheless, as `go mod` will eventually cease to be experimental, it is worth considering the implications.

# Expected result

- Allow developers to use either `dep` or `go mod`, at least until the baseline supported version in kata is above 1.11. Consider removing `dep` support afterwards.
- Remove `vendor` from the repository.

# Actual result

- Only `dep` is supported.
- The `vendor` repository is checked in the repo and it must be explicitly updated. See [#1152](https://github.com/kata-containers/runtime/pull/1152), [community#88](https://github.com/kata-containers/community/issues/88) and [community#89](https://github.com/kata-containers/community/issues/89).

---

So, regarding

> [@grahamwhaley]
> Now, if you'd like to propose a move to `go mod`, and submit a PR ;-) But, first probably best if it was discussed.

> [@sboeuf]
> @umarcor is it something you're interested in helping with?

I said I was asking because of curiosity since I don't really understand how `go mod` is expected to work. I see that it always works in clean environments (such as CI jobs), but when used in a machine with other golang packages, it freaks out easily. Precisely, the behaviour that I have seen is the following:

- If a dependency does not exist locally, the version specified in `go.mod` is retrieved.
- If the dependency exists locally, it may work or it may complain because the version does not match.
  - This is sometimes desirable, because the local dependency is in a valid dirty state, i.e. some patches have been applied.
  - In other contexts, the checked out or modified local state corresponds to any other project, and it is not compatible, so it breaks the build.

Moreover, `go mod` allows to generate a `vendor` directory, but it is not honoured. I.e., I would expect it to search for dependencies in common paths (`/go`, `~/go`, `GOPATH`) and save additional clones in the `vendor` only when there is some conflict. Then, when `go build` or `go run` is executed, it would use the versions in `vendor` and, if not present, it would search in the common paths.

What is happening instead is that you can copy absolutely all the dependencies to `vendor` and then tell go tools to use it exclusively. This is quite nonesense, because it effectively treats each repository as a completely independent project and the same dependencies are duplicated unnecessarily. At this point, I don't know if there is any benefit over using git submodules, provided that all the go dependencies are available as git repos.

Therefore, an 'easy' workaround is to just use a `golang:alpine` container and build the binaries with `CGO_ENABLED=0 go build -a`. This allows to execute each build in a clean environment, so there is no conflict with other projects in the host. The main disadvantage of this approach is that all the dependencies are downloaded each time a container is started, unless `vendor` is used exclusively.

Ideally, checking `vendor` into the repo should not be required neither with `dep` nor with `go mod`. The `vendor` directory should be a local temporal resource that can be 'generated' unequivocally with `go.mod`, `Gopkg.lock` and/or `Gopkg.toml`.

Last, the behaviour of `go mod` is different if the repository is inside or outside of the `GOTPATH`.

---

Anyway, the 'clean' approach to migrate from `dep` to `go mod` should be as 'simple' as:

``` bash
docker run --rm -it golang:alpine sh -c "$(cat <<-EOF
# Get 'regular' dependencies to build the runtime
apk update
apk add git
go get -u github.com/golang/dep/cmd/dep
go get -u github.com/kata-containers/runtime
cd \$GOPATH/src/github.com/kata-containers/runtime
dep ensure

# Remove 'dep' related sources
git rm -rf Gopkg.lock Gopkg.toml vendor/

# Move the repo outside of GOPATH
cd ..
mv runtime ~/
cd ~/runtime

# Initialize go modules
go mod init
git add go.mod

# Profit
git config --local user.name "dep2mod"
git config --local user.email "dep2mod@script.example"
git status
git commit -m 'migrate from dep to go mod'
git push
EOF
)"
```

> NOTE: optionally, replace `git push` with `sh` to execute further commands before pushing the result.

> NOTE: on MSYS2 (Win10), prepend the first line with `winpty`.

---

Summarizing, @grahamwhaley, @sboeuf, I am interested on the discusion, on hearing your thoughs, and on seen how a project like this handles this specific issue (provided that dependency management in golang is a flaw in the language/ecosystem). Nevertheless, I can sure open PRs applying the script above to the repos you deem appropriate.

### PR DESCRIPTION
**Pull request by amshinde**
Move from go dep to go modules.

Fixes #1331

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -626,6 +626,12 @@ install-completions:
 clean:
 	$(QUIET_CLEAN)rm -f $(TARGET) $(SHIMV2) $(NETMON_TARGET) $(CONFIGS) $(GENERATED_FILES) .git-commit .git-commit.tmp
 
+vendor:
+	export GO111MODULE=on \
+	go mod tidy && \
+	go mod vendor && \
+	go mod verify
```
Comment (jodh-intel) [inline Makefile]: Nice. Could you add a check like this to https://github.com/kata-containers/tests/blob/master/.ci/static-checks.sh for repos that are using go modules? Yes, I know the check is already here, but having such a central and PR-gating check will hopefully keep all the repos protected.

**Comment by ganeshmaharaj** [comment]
@amshinde Is the plan to get runtime moved while we move rest of the repos? In that case, tests repository needs to be bostered to run `go mod verify` against `dep check` for just these repos.

**Comment by jodh-intel** [comment]
@amshinde, @ganeshmaharaj - please can we document the plan on https://github.com/kata-containers/kata-containers/issues/42 to allow others to understand (and comment on) it.

File: Makefile (hunk 1 above)
Comment (amshinde) [reply Makefile]: @jodh-intel I have finally addressed this after my unexpected interim break.
I have opened a PR in the tests repo to account for go modules.
See: https://github.com/kata-containers/tests/pull/1880

**Comment by amshinde** [comment]
@jodh-intel Will update the issue.

File: Makefile (hunk 1 above)
Comment (jodh-intel) [reply Makefile]: Thanks @amshinde.

**Comment by caoruidong** [comment]
ping @amshinde

**Comment by umarcor** [comment]
Hi @amshinde, @jodh-intel! Coming from #1420, I'm sorry to be the one bringing bad news here...

Context:

- Currently, this repository retrieves the tests repo with 'go get' and expects it to be available in `GOPATH/src/github.com/kata-containers/tests`. However, in any environment where go modules are enabled, `go get` will NOT download sources of packages in that path. This is why this PR will fail when go modules are enabled in Travis: https://travis-ci.com/umarcor/kata-runtime/jobs/231319730. See https://github.com/kata-containers/runtime/pull/1420#issuecomment-528176791.
- With go modules, 'vendor' is expected not to be required and there is no idiomatic way to properly 'mix' sources available in the modules caches and those in 'vendor'. This is related to https://github.com/kata-containers/tests/pull/1880#issue-303163610.
  - Any build with go modules should be reproducible, as long as remotes are not gone. This is the only use case of 'vendor' in the new environment: a backup, just in case some of the remotes fails. As a result, 'vendor' should not be included in the VCS. In any case, vendor packages could be distributed as tarballs for users with very specific requirements to have them.
  - By the  same token, build commands need to be specifically forced to use the vendor folder.

Therefore, I think that this PR  should ensure that go modules are not used, except for running `go mod vendor`.

**Comment by WeiZhang555** [comment]
I think `go mod vendor` can be an option. How about we only use `go mod` to manage our "vendor" dir instead of remove whole "vendor" from the repo?

**Comment by umarcor** [comment]
> I think `go mod vendor` can be an option. How about we only use `go mod` to manage our "vendor" dir instead of remove whole "vendor" from the repo?

That's a temporal solution that will only work as long as go modules are optional. Nevertheless, as commented, this approach would require:

- *this PR should ensure that go modules are not used, except for running `go mod vendor`*.
- *build commands need to be specifically forced to use the vendor folder*.

**Comment by jcvenegas** [comment]
@amshinde ping this needs rebase

**Comment by raravena80** [comment]
@amshinde any updates on this PR? Thx!

**Comment by HZ89** [comment]
@amshinde any updates on this PR? Thx!

**Comment by amshinde** [comment]
I'll revisit this PR this week.

**Comment by raravena80** [comment]
@amshinde any updates? Thx.

Your weekly Kata herder.

**Comment by raravena80** [comment]
@amshinde ping 😄

**Comment by bergwolf** [comment]
> Currently, this repository retrieves the tests repo with 'go get' and expects it to be available in GOPATH/src/github.com/kata-containers/tests. However, in any environment where go modules are enabled, go get will NOT download sources of packages in that path. This is why this PR will fail when go modules are enabled in Travis: https://travis-ci.com/umarcor/kata-runtime/jobs/231319730. See #1420 (comment).

We should not use the **new** `go get` to download code or binary inside a go module directory, -- it would fetch the code and add it to your `go.mod` dependency. `git clone` is your friend ;)

And kindly ping @amshinde from your Kata Herder this week ;)

**Comment by umarcor** [comment]
> We should not use the **new** `go get` to download code or binary inside a go module directory, -- it would fetch the code and add it to your `go.mod` dependency. `git clone` is your friend ;)

@bergwolf, thanks for acknowledging this point. I've trying to make it clear for almost a year now.. #1420, #1413. I believe that's the critical issue that is preventing any practical movement towards go modules.

**Comment by raravena80** [comment]
@amshinde @bergwolf @umarcor ping.

**Comment by umarcor** [comment]
See https://github.com/kata-containers/runtime/pull/1420#issuecomment-700777378.

**Pull request closed without merging**